Documenting how to use EvenPodsSpread feature

Huang-Wei opened the pull request:
Here is a preview of the document: https://deploy-preview-15514--kubernetes-io-vnext-staging.netlify.com/docs/concepts/workloads/pods/pod-topology-spread-constraints/

cc/ @bsalamat @ahg-g @alculquicondor
/sig scheduling
/milestone v1.16

daminisatya commented:
@Huang-Wei I am the 1.16 Docs Shadow. We need this PR ready for review before Sep 3.

Huang-Wei commented:
@bsalamat @ahg-g @alculquicondor The PR is ready for review. We will focus on the technical part. And after that, I will request an editorial review.

Huang-Wei commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,265 @@
+You can define one or multiple `topologySpreadConstraint` to instruct the kube-scheduler to how to place the incoming pod with existing pods across the cluster. In particular, the fields are explained as below:
+
+- **maxSkew** describes the degree to which pods may be unevenly distributed. It's the maximum permitted difference between the number of matching pods in any two topology domains of a given topology type.
+- **topologyKey** is the key of node labels. Nodes that have a label with this key and identical values are considered to be in the same topology. We consider each <key, value> as a "bucket", and try to put a balanced number of pods into each bucket.
```
That is mentioned in other sections.

Huang-Wei commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,265 @@
+- **whenUnsatisfiable** indicates how to deal with a pod if it doesn't satisfy the spread constraint:
+    - `DoNotSchedule` (default) tells the scheduler not to schedule it
+    - `ScheduleAnyway` tells the scheduler to still schedule it
+- **labelSelector** is used to find matching pods. Pods that match this label selector are counted to determine the number of pods in their corresponding topology domain.
```
SG.

Huang-Wei commented:
@alculquicondor Appreciate a lot for your review! Comments have been addressed.

alculquicondor commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -210,50 +204,59 @@ If we apply "mypod.yaml" to this cluster, you will notice "mypod" stays in `Pend
 To overcome this situation, you can either increase the `maxSkew` or modify one of the constraints to use `whenUnsatisfiable: ScheduleAnyway`.
 
-### Example: Respecting NodeSelecor and NodeAffinity
+### Conventions
```
I think this section should go before the examples.

Huang-Wei replied:
I'm ok either way. My original thought is that we should show some examples to give a general idea on this feature, and then explain some underlying conventions worthing noting.

alculquicondor replied:
sg

alculquicondor commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -210,50 +204,59 @@ If we apply "mypod.yaml" to this cluster, you will notice "mypod" stays in `Pend
-            - zoneC
-...
-```
+- Be aware of what happends if the incomingPod’s `topologySpreadConstraints[*].labelSelector` doesn’t match its own labels. In the above example, if we remove the incoming pod’s labels, it can still be placed onto "zoneB" since the constraints are respected well. However, after the placement, the degree of imbalance of the cluster remains unchanged - it’s still zoneA having 2 pods which hold labe ...
````
s/happends/happens

alculquicondor replied:
just this from my end

alculquicondor commented:
/lgtm

ahg-g commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,268 @@
+- **topologyKey** is the key of node labels. Nodes that have a label with this key and identical values are considered to be in the same topology. We consider each <key, value> as a "bucket", and try to put a balanced number of pods into each bucket.
+- **whenUnsatisfiable** indicates how to deal with a pod if it doesn't satisfy the spread constraint:
+    - `DoNotSchedule` (default) tells the scheduler not to schedule it.
+    - `ScheduleAnyway` tells the scheduler to still schedule it in the manner of prioritizing nodes that minimize the skew.
```
s/in the manner of/while

ahg-g commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,268 @@
+
+### Example: One TopologySpreadConstraint
+
+Suppose we have a 4-node cluster where 3 pods labeled `foo:bar` are located from node1 to node3 respectively (`P` represents Pod):
```
s/from node1 to node3 respectively/in node1, node2 and node3

ahg-g commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,268 @@
++-------+-------+-------+
+```
+
+If we apply "mypod.yaml" to this cluster, you will notice "mypod" stays in `Pending` state. This is because: to satisfy the first constraint, "mypod" can only be put to "zoneB"; while in terms of the second constraint, "mypod" can only put to "node2". Then a joint result of "zoneB" and "node2" returns nothing.
````
"mypod.yaml" as a file is not defined before, so perhaps say: "If we try to create _mypod_ as defined above"

Huang-Wei replied:
I will include it as a standalone .yaml file later.

ahg-g commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,268 @@
+
+- Only the pods holding the same namespace as the incoming pod can be matching candidates.
+
+- Nodes without `topologySpreadConstraints[*].topologyKey` present will be bypassed - i.e. the pods located on those nodes being disregarded. In the above example, suppose a node5 carrying label `{zone-typo: zoneC}` joins the cluster, it will be bypassed due to the absence of label key `zone`.
```
s/i.e. the pods located on those nodes being disregarded/i.e., pods scheduled on those nodes do not impact maxSkew calculation

ahg-g replied:
when you say bypassed, does this also mean that the node will be filtered (i.e., the incoming pod will not be scheduled on it), if yes we should clarify that as well?

Huang-Wei replied:
You're right, it means both (1) pods on those nodes won't impact maxSkew calculation, and (2) the incoming pod won't be scheduled onto this kind of nodes.

ahg-g commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,268 @@
+
+- Nodes without `topologySpreadConstraints[*].topologyKey` present will be bypassed - i.e. the pods located on those nodes being disregarded. In the above example, suppose a node5 carrying label `{zone-typo: zoneC}` joins the cluster, it will be bypassed due to the absence of label key `zone`.
+
+- Be aware of what happends if the incomingPod’s `topologySpreadConstraints[*].labelSelector` doesn’t match its own labels. In the above example, if we remove the incoming pod’s labels, it can still be placed onto "zoneB" since the constraints are respected well. However, after the placement, the degree of imbalance of the cluster remains unchanged - it’s still zoneA having 2 pods which hold labe ...
```
s/respected well/still satisfied

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+## EvenPodsSpread
+
+`EvenPodsSpread` feature gives users more fine-grained control on the pods' distribution across the cluster among failure-domains such as regions, zones, nodes, and other user-defined topology domains. This can help to achieve high availability as well as efficient resource utilization.
```
Some of this belongs in the page overview. When revising, consider these tweaks:
```suggestion
To use _spread constraints_, you need to enable the `EvenPodsSpread` https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/).
```

Huang-Wei replied:
We can simply remove this section as the content has been moved to the "overview" section.

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+---
+title: Even Pods Spread
```
```suggestion
title: Pod spread constraints
```
The feature gate is EvenPodsSpread; the _concept_ is Pod spread. Have I got that right?

sftim replied:
I'd expect to find this page inside https://kubernetes.io/docs/concepts/workloads/pods/

Huang-Wei replied:
How about "Pod topology spread constraints"?

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+## Motivation
+
+In Kubernetes, "Affinity" related directives are aimed to control how pods are
```
```suggestion
In Kubernetes, directives related to “Affinity” control how Pods are
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+## Motivation
+
+In Kubernetes, "Affinity" related directives are aimed to control how pods are
+scheduled - more packed or more scattered. But right now only limited options
```
[Avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) as per the style guide.

Huang-Wei replied:
+100 on this! Didn't know we have this guide.

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+In Kubernetes, "Affinity" related directives are aimed to control how pods are
+scheduled - more packed or more scattered. But right now only limited options
+are offered: for `PodAffinity`, infinite pods can be packed into qualifying
```
```suggestion
 are offered: for `PodAffinity`, you can try to pack any number of Pods into qualifying
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+topology domains - to achieve high availability or cost-saving. And
+regular rolling upgrade or scaling out replicas can also be problematic.
+
+## Set up
```
I recommend switching this Set up section into a new Task page.

Huang-Wei replied:
Hmm.. Actually, this is not a serious "set up" section. Maybe reword to "Prerequisites"?

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
+This feature requires the nodes to be labelled with topological key-value pairs via `kubectl label node <node_name> <key1=val1> <key2=val2> ...`. For example, the following 4-node cluster is tagged with {node: node_name>, zone: zone_name}:
+
+```
+$ kubectl get node --show-labels
````
Per the style guide:
- avoid putting the prompt into shell snippets
- separate the command from its sample output

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
+
+So the cluster is logically viewed as below:
+
+```
````
Would it help to have a proper diagram here? I might be able to help draw that.

Huang-Wei replied:
A diagram _can_ help. I'm just thinking of text styled diagram, which should be enough and portable.

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
++-------+-------+-------+-------+
+```
+
+Instead of manually applying labels, you can also reuse the [standardized Kubernetes labels](/docs/reference/kubernetes-api/labels-annotations-taints/) which are created and populated automatically by most cluster providers.
````
```suggestion
Instead of manually applying labels, you can also reuse the [well-known labels](/docs/reference/kubernetes-api/labels-annotations-taints/) that are created and populated automatically on most clusters.
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+### API
+
+A new `pod.spec.topologySpreadConstraints` API is introduced as below:
```
Similar to [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future), be cautious about making statements about the present.

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+### Example: One TopologySpreadConstraint
+
+Suppose we have a 4-node cluster where 3 pods labeled `foo:bar` are located in node1, node2 and node3 respectively (`P` represents Pod):
```
```suggestion
Suppose you have a 4-node cluster where 3 pods labeled `foo:bar` are located in node1, node2 and node3 respectively (`P` represents Pod):
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+`topologyKey: zone` implies the even distribution will only be applied to the nodes which have label pair "zone:<any value>" present.  And `whenUnsatisfiable: DoNotSchedule` tells the scheduler to let it stay pending if the incoming pod can’t satisfy the constraint.
+
+If we place the pod onto "zoneA", the pods distribution would become [3, 1], hence the actual skew is 2 (3 - 1) which violates `maxSkew: 1`, so in this example, the incoming pod can only be placed onto "zoneB":
```
```suggestion
If the control plane places the incoming Pod onto "zoneA", the pods distribution would become [3, 1], hence the actual skew is 2 (3 - 1) which violates `maxSkew: 1`. In this example, the incoming pod can only be placed onto "zoneB":
```

Huang-Wei replied:
We can just say:

> If the Pod is placed into "zoneA"

BTW: "onto", or "into"? Any suggestion?

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
++-------+-------+-------+-------+      +-------+-------+-------+-------+
+```
+
+We can tweak the pod spec to meet various kinds of requirements:
````
```suggestion
You can tweak the pod spec to meet various kinds of requirements:
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+### Example: Multiple TopologySpreadConstraints
+
+Let’s reuse the above example: suppose we have a 4-node cluster where 3 pods labeled `foo:bar` are located from node1 to node3 respectively (`P` represents Pod):
```
```suggestion
This builds upon the previous example. Suppose you have a 4-node cluster where 3 pods labeled `foo:bar` are located from node1 to node3 respectively (`P` represents Pod):
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
++-------+-------+-------+-------+
+```
+
+We can use 2 TopologySpreadConstraints to control the pods spreading on both zone and node:
````
```suggestion
You can use 2 TopologySpreadConstraints to control the pods spreading on both zone and node:
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
+...
+```
+
+In this case, to match the first constraint, the incoming pod can only be placed onto "zoneB"; while in terms of the second constraint, the incoming pod can only be placed onto "node4". Then the results of 2 constraints are ANDed, so we get the only option "node4".
````
```suggestion
In this case, to match the first constraint, the incoming pod can only be placed onto "zoneB"; while in terms of the second constraint, the incoming pod can only be placed onto "node4". Then the results of 2 constraints are ANDed, so the only viable option is to place on "node4".
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+In this case, to match the first constraint, the incoming pod can only be placed onto "zoneB"; while in terms of the second constraint, the incoming pod can only be placed onto "node4". Then the results of 2 constraints are ANDed, so we get the only option "node4".
+
+Be aware that multiple constraints can sometimes lead to conflicts. Suppose we have a 3-node cluster across 2 zones:
```
```suggestion
Multiple constraints can lead to conflicts. Suppose you have a 3-node cluster across 2 zones:
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
++-------+-------+-------+
+```
+
+If we apply "mypod.yaml" to this cluster, you will notice "mypod" stays in `Pending` state. This is because: to satisfy the first constraint, "mypod" can only be put to "zoneB"; while in terms of the second constraint, "mypod" can only put to "node2". Then a joint result of "zoneB" and "node2" returns nothing.
````
```suggestion
If you apply "mypod.yaml" to this cluster, you will notice "mypod" stays in `Pending` state. This is because: to satisfy the first constraint, "mypod" can only be put to "zoneB"; while in terms of the second constraint, "mypod" can only put to "node2". Then a joint result of "zoneB" and "node2" returns nothing.
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+- If the incoming pod has `spec.nodeSelector` or `spec.affinity.nodeAffinity` defined, nodes not matching them will be bypassed.
+
+    Suppose we have a 5-node cluster ranging from zoneA to zoneC:
```
```suggestion
    Suppose you have a 5-node cluster ranging from zoneA to zoneC:
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
+    +-------+-------+-------+-------+-------+
+    ```
+
+    If we’re clear that zoneC must be excluded, we can compose the spec like this:
````
```suggestion
    and you know that that zoneC must be excluded. In that case, you can compose the spec like this:
```

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
+
+    If we’re clear that zoneC must be excluded, we can compose the spec like this:
+
+    ```
````
nit: Mark this as YAML

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
````diff
@@ -0,0 +1,270 @@
+
+We can use 2 TopologySpreadConstraints to control the pods spreading on both zone and node:
+
+```
````
nit: Mark this as YAML

sftim commented on content/en/docs/concepts/configuration/even-pods-spread.md:
```diff
@@ -0,0 +1,270 @@
+
+{{< feature-state for_k8s_version="v1.16" state="alpha" >}}
+
+This page explains how EvenPodsSpread feature works and the mechanics under the hood.
```
```suggestion
You can use _spread constraints_ to control how Pods are spread across your cluster among failure-domains such as regions, zones, nodes, and other user-defined topology domains. This can help to achieve high availability as well as efficient resource utilization.
```

sftim replied:
Also consider using a glossary tooltip for Pods: `{{< glossary_tooltip text="Pods" term_id="pod" >}}`

Huang-Wei replied:
It's more precise to use "topology spread constraints".

Huang-Wei commented:
@sftim Thanks a lot for the review. The comments have been addressed. PTAL.

for the review's convenience, I intentionally made the change of file renaming to a separate commit.

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+## Motivation
+
+In Kubernetes, directives related to "Affinity" control how Pods are
+scheduled - more packed or more scattered. But right now only limited options
```
As soon as this is merged into the _master_ branch, “right now” will be incorrect: there will be a new (alpha) feature that provides an alternative.

sftim replied:
I recommend moving this Motivation section further down the page, or even omitting it and putting a link into the What's Next section so that interested readers can find and read the KEP.

Huang-Wei replied:
How about moving it down the page, and rename to "Comparison with PodAffinity/PodAntiAffinity".

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+### Enable Feature Gate
+
+Ensure the `EvenPodsSpread` feature gate is enabled (it is disabled by default
+in 1.16). See [Feature Gates](/docs/reference/command-line-tools-reference/feature-gates/) for an explanation of enabling feature gates. The `EvenPodsSpread` feature gate must be enabled on apiserver **and** scheduler.
```
```suggestion
in 1.16). See [Feature Gates](/docs/reference/command-line-tools-reference/feature-gates/) for an explanation of enabling feature gates. The `EvenPodsSpread` feature gate must be enabled for the API server **and** for the scheduler.
```

Consider using glossary tooltips for “API server” and “scheduler”.

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+topology domain(s); for `PodAntiAffinity`, only one Pod can be scheduled into a
+single topology domain.
+
+This is not an ideal situation if users want to distribute Pods evenly across different
```
If keeping this section:
```suggestion
This is not ideal if you want to distribute Pods evenly across different
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+Ensure the `EvenPodsSpread` feature gate is enabled (it is disabled by default
+in 1.16). See [Feature Gates](/docs/reference/command-line-tools-reference/feature-gates/) for an explanation of enabling feature gates. The `EvenPodsSpread` feature gate must be enabled on apiserver **and** scheduler.
+
+### Apply Labels to Worker Nodes
```
This line reads like part of a Task description; however, it's in a page that explains a concept.

How about:
```suggestion
### Node labels
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+
+### Apply Labels to Worker Nodes
+
+This feature requires the nodes to be labelled with topological key-value pairs via `kubectl label node <node_name> <key1=val1> <key2=val2> ...`. For example, you can verify using `--show-labels` flag:
```
```suggestion
Topology spread constraints rely on node labels to identify the topology domain(s) that each Node is in. For example, a Node might have labels: `node=node1,zone=af-northeast-1b,region=africa-northeast`
```

In a Concept guide, the docs don't need to explain how to set and get Node labels.

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
````diff
@@ -0,0 +1,271 @@
++-------+-------+-------+-------+
+```
+
+Instead of manually applying labels, you can also reuse the [well-known labels](/docs/reference/kubernetes-api/labels-annotations-taints/) that are created and populated automatically on most clusters.
````
:ok_hand:

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+
+Instead of manually applying labels, you can also reuse the [well-known labels](/docs/reference/kubernetes-api/labels-annotations-taints/) that are created and populated automatically on most clusters.
+
+## Usage
```
```suggestion
## Spread constraints for Pods
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
````diff
@@ -0,0 +1,271 @@
+    labelSelector: <object>
+```
+
+You can define one or multiple `topologySpreadConstraint` to instruct the kube-scheduler how to place the incoming Pod with existing Pods across the cluster. In particular, the fields are explained as below:
````
```suggestion
You can define one or multiple `topologySpreadConstraint` to instruct the kube-scheduler how to place each incoming Pod in relation to the existing Pods across your cluster. The fields are:
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+You can define one or multiple `topologySpreadConstraint` to instruct the kube-scheduler how to place the incoming Pod with existing Pods across the cluster. In particular, the fields are explained as below:
+
+- **maxSkew** describes the degree to which Pods may be unevenly distributed. It's the maximum permitted difference between the number of matching Pods in any two topology domains of a given topology type. It must be greater than zero.
+- **topologyKey** is the key of node labels. Nodes that have a label with this key and identical values are considered to be in the same topology. We consider each <key, value> as a "bucket", and try to put a balanced number of Pods into each bucket.
```
Is this easier to read (and still correct technically)?
```suggestion
- **topologyKey** is the key of node labels. If two Nodes are labelled with this key and have identical values for that label, the scheduler treats both Nodes as being in the same topology. The scheduler tries to place a balanced number of Pods into each topology domain.
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+- **whenUnsatisfiable** indicates how to deal with a Pod if it doesn't satisfy the spread constraint:
+    - `DoNotSchedule` (default) tells the scheduler not to schedule it.
+    - `ScheduleAnyway` tells the scheduler to still schedule it while prioritizing nodes that minimize the skew.
+- **labelSelector** is used to find matching Pods. Pods that match this label selector are counted to determine the number of Pods in their corresponding topology domain. See [Lable Selectors](/docs/concepts/overview/working-with-objects/labels/#label-selectors) for more details.
```
nit:
```suggestion
- **labelSelector** is used to find matching Pods. Pods that match this label selector are counted to determine the number of Pods in their corresponding topology domain. See [Label Selectors](/docs/concepts/overview/working-with-objects/labels/#label-selectors) for more details.
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+    - `ScheduleAnyway` tells the scheduler to still schedule it while prioritizing nodes that minimize the skew.
+- **labelSelector** is used to find matching Pods. Pods that match this label selector are counted to determine the number of Pods in their corresponding topology domain. See [Lable Selectors](/docs/concepts/overview/working-with-objects/labels/#label-selectors) for more details.
+
+Detailed API explanation can also be found via `kubectl explain Pod.spec.topologySpreadConstraints`.
```
nit:
```suggestion
You can read more about this field by running `kubectl explain Pod.spec.topologySpreadConstraints`.
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
````diff
@@ -0,0 +1,271 @@
+...
+```
+
+`topologyKey: zone` implies the even distribution will only be applied to the nodes which have label pair "zone:<any value>" present.  And `whenUnsatisfiable: DoNotSchedule` tells the scheduler to let it stay pending if the incoming Pod can’t satisfy the constraint.
````
nit:
```suggestion
`topologyKey: zone` implies the even distribution will only be applied to the nodes which have label pair "zone:<any value>" present. `whenUnsatisfiable: DoNotSchedule` tells the scheduler to let it stay pending if the incoming Pod can’t satisfy the constraint.
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+
+`topologyKey: zone` implies the even distribution will only be applied to the nodes which have label pair "zone:<any value>" present.  And `whenUnsatisfiable: DoNotSchedule` tells the scheduler to let it stay pending if the incoming Pod can’t satisfy the constraint.
+
+If the Pod is placed into "zoneA", the Pods distribution would become [3, 1], hence the actual skew is 2 (3 - 1) which violates `maxSkew: 1`. In this example, the incoming Pod can only be placed onto "zoneB":
```
```suggestion
If the scheduler placed this incoming Pod into "zoneA", the Pods distribution would become [3, 1], hence the actual skew is 2 (3 - 1) which violates `maxSkew: 1`. In this example, the incoming Pod can only be placed onto "zoneB":
```

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+
+There are some implicit conventions worth noting here:
+
+- Only the Pods holding the same namespace as the incoming Pod can be matching candidates.
```
This seems important, and I recommend mentioning it earlier. It's OK to repeat this so that it's clear to readers.

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -0,0 +1,271 @@
+- Change `topologyKey` to "node" so as to distribute the Pods evenly across nodes instead of zones. In the above example, if `maxSkew` remains "1", the incoming Pod can only be placed onto "node4".
+- Change `whenUnsatisfiable: DoNotSchedule` to `whenUnsatisfiable: ScheduleAnyway` to ensure the incoming Pod to be always schedulable (suppose other scheduling APIs are satisfied). However, it’s preferred to be placed onto the topology domain which has fewer matching Pods. (Be aware that this preferability is jointly normalized with other internal scheduling priorities like resource usage ratio, ...
+
+### Example: Multiple TopologySpreadConstraints
```
Consider moving this to a new Task page, eg one titled “Configure topology spread constraints for a workload”.

That will help readers find relevant Task-focued documentation if that's what they're after, whilst keeping the Concept documentation focused.

Huang-Wei commented:
@sftim I've addressed most of your comments. Some of them were kept as is. Can you take another review? We may need a /lgtm by the end of today.

Thanks.

sftim commented:
@Huang-Wei you can help reviewers by putting a link to https://deploy-preview-15514--kubernetes-io-vnext-staging.netlify.com/docs/concepts/workloads/pods/pod-topology-spread-constraints/#prerequisites in the PR description.

Informally (this isn't a tech review), it /lgtm

I still think it's worth tracking an issue to replace the ASCII-art style diagrams with SVG ones (and I'm fine with the idea of merging this now without those diagrams done).

ahg-g commented:
/lgtm

sftim commented:
@Huang-Wei I sent a PR against your branch: https://github.com/Huang-Wei/website/pull/1

This adds an optional inbound link to the new concept page.

Huang-Wei commented:
Thanks @sftim . I've merged your commit.

ahg-g commented:
/lgtm

simplytunde commented:
/approve
